resources/testdriver-actions.js

```diff
@@ -0,0 +1,383 @@
+(function() {
+  let sourceNameIdx = 0;
+
+  /**
+   * Builder for creating a sequence of actions
+   */
+  function Actions() {
+    this.sourceTypes = new Map([["key", KeySource],
+                                ["pointer", PointerSource],
+                                ["general", GeneralSource]]);
+    this.sources = new Map();
+    for (let sourceType of this.sourceTypes.keys()) {
+      this.sources.set(sourceType, new Map());
+    }
+    this.currentSources = new Map();
+    for (let sourceType of this.sourceTypes.keys()) {
+      this.currentSources.set(sourceType, null);
+    }
+    this.createSource("general");
+    this.tickIdx = 0;
+  }
+
+  Actions.prototype = {
+    /**
+     * Generate the action sequence suitable for passing to
+     * test_driver.action_sequence
+     *
+     * @returns {Array} Array of WebDriver-compatible actions sequences
+     */
+    serialize: function() {
+      let actions = [];
+      for (let [_, sources] of this.sources) {
+        for (let [sourceName, source] of sources) {
```

Comment:
Based on what I understood the way we store/serialized these actions is basically a few parallel lists. I was wondering why we chose that. Correct me if I'm wrong. But in one tick we might have key/pointer actions from different names. Do you leave it to the implementation to choose which one to inject first? Because I don't think we can inject anything at literally the same time. We can inject them back to back and in the same frame but not at the same time. Also this comes into play as one of the actions in one tick could be pause and the question is whether we should pause before injecting other actions in that tick or after.

Reply:
The spec processes events in a specific order; basically in the order the input devices are added (we work top to bottom on the per-device lists of events). Honestly I'm not sure why we didn't just have a single list rather than multiple parallel lists; it seems like a bad design to me (and I"m at least particually responsible), but it is what it is.

Reply:
I feel like this will bite us down the road when drivers and possibly the test-driver-vendor.js (which could override how this command is routed/executed by vendors) treat this ordering differently. But again I feel like this multi source things at the same time is not a common usecase and definitely not worth blocking this PR for.

Reply:
I mean the deal here is clearly that the semantics are defined in the WebDriver spec, and that does specify an order. I guess we can allow only one action per tick. Although I should have pointed out that pause doesn't work like that; the pause is asynchronous from the point of view of the rest of the actions in the tick. The algorithm is something like

```
let tickEnd;
if (tick.duration) {
   tickEnd = new Promise(resolve => setTimeout(resolve, duration));
} else {
   tickEnd = new Promise(resolve => resolve())
}
for (action of tick.actions) {
  action.dispatch();
}
await tickEnd;
```